Formula/tremor-runtime.rb

```diff
@@ -0,0 +1,61 @@
+class TremorRuntime < Formula
+  desc "Early-stage event processing system for unstructured data"
+  homepage "https://www.tremor.rs/"
+  url "https://github.com/tremor-rs/tremor-runtime/archive/refs/tags/v0.12.4.tar.gz"
+  sha256 "91cbe0ca5c4adda14b8456652dfaa148df9878e09dd65ac6988bb781e3df52af"
+  license "Apache-2.0"
+  head "https://github.com/tremor-rs/tremor-runtime.git", branch: "main"
+
+  bottle do
+    root_url "https://github.com/sandipb/homebrew-tremor/releases/download/tremor-runtime-0.12.4"
+    sha256 cellar: :any_skip_relocation, monterey: "eccbc7e61188daee1e2deadf04752d5662f05ded2e68741b431a8a9523beb74a"
+    sha256 cellar: :any_skip_relocation, big_sur:  "4de4e40f728af2c1b17aef831e66a59177a8be25e93a8a8c68951d88fd27d438"
+  end
+
+  depends_on "cmake" => :build
+  depends_on "rust" => :build
+
+  def install
+    # To fix SIGILL. https://github.com/tremor-rs/tremor-runtime/issues/1221
+    ENV["RUSTFLAGS"] = "-C target-cpu=native"
```

Comment:
I don't think this is a good idea.

This looks like this will create a non-portable binary, because it will generate machine instructions that are not available on CPUs older than the ones in our build machines.

Reply:
According to the issue I documented in the comment, the problem is that the code tries to use advanced machine instructions that may not be available in the build workers.

The build fails without this. 

What would you suggest?

Reply:
What are the possible values for `target-cpu`?

Reply:
The official doc is here: https://rust-lang.github.io/packed_simd/perf-guide/target-feature/rustflags.html#target-cpu

I think the flag puts an upper bound for cpu features. *not* using the flag causes it to generate instructions that dont even work on the build machines.

We are already building this app on atleast 3 (Mac, intel, arm) + 2 (Linux) different arches as per the documented requirements of formulas. Are we targeting other CPUs as well? Because of the requirements, at my end, I am already testing this on mac and linux to troubleshoot the builds. I cannot possibly test on other architectures on my own.

Is this going to be a blocker for this formula? I already spent all night fixing the other (excellent, I must emphasize) enhancements that you suggested.

But if I need to test it on other arches to get accepted, I would probably stop here. I know you are trying to do the best for the users, but given that a whole range of arches are already working, can't we work on a fix later if someone really has a problem?

Reply:
> I think the flag puts an upper bound for cpu features.

I'm afraid you're mistaken. From the issue you linked:

> It won't make sense for packages as then everyone is bound to the CPU the CI machine runs

And from the documentation you linked:

> Using `native` as the CPU model will cause Rust to generate and optimize code for the CPU running the compiler. It is useful when building programs which you plan to only use locally. This should never be used when the generated programs are meant to be run on other computers, such as when packaging for distribution or cross-compiling.

We are packaging this software for distribution, so we definitely _should not_ be using `target-cpu=native` here.

> _not_ using the flag causes it to generate instructions that dont even work on the build machines.

This is because of your `.cargo/config`:
```toml
[build]
rustflags = ["-C", "target-feature=+avx,+avx2,+sse4.2"]
```

That said, I think our CI actually supports all these instruction sets, so dropping the `native` flag actually probably has no effect on CI but _will_ cause issues for some users.

> We are already building this app on atleast 3 (Mac, intel, arm) + 2 (Linux) different arches as per the documented requirements of formulas. Are we targeting other CPUs as well? Because of the requirements, at my end, I am already testing this on mac and linux to troubleshoot the builds. I cannot possibly test on other architectures on my own.
> 
> Is this going to be a blocker for this formula? I already spent all night fixing the other (excellent, I must emphasize) enhancements that you suggested.
> 
> But if I need to test it on other arches to get accepted, I would probably stop here. I know you are trying to do the best for the users, but given that a whole range of arches are already working, can't we work on a fix later if someone really has a problem?

Note that `target-cpu` isn't about target architectures, but about target microarchitectures, so this isn't about considering machines that aren't running macOS or Linux on Intel or ARM.

Here's what this should say:

```suggestion
    cpu = case Hardware.oldest_cpu
    when :arm_vortex_temptest then "apple-m1"
    else Hardware.oldest_cpu
    end
    ENV["RUSTFLAGS"] = "-C target-cpu=#{cpu}"
```

Alternatively, this likely also works:

```suggestion
    ENV["RUSTFLAGS"] = "-C target-cpu=#{Hardware.oldest_cpu}" if Hardware::CPU.intel?
```

Reply:
Big respect to your experienced take on this. Sorry for the 5 a.m. rant. :)  Will try this out.

Reply:
@carlocab Seems like this app will not work on cpu which doesn't support simd.

```
     --> /github/home/.cache/Homebrew/cargo_cache/registry/src/github.com-1ecc6299db9ec823/simd-json-0.5.0/src/lib.rs:221:86
      |
  221 | fn please_compile_with_a_simd_compatible_cpu_setting_read_the_simdjsonrs_readme() -> ! {}
      |    ----------------------------------------------------------------------------      ^ expected `!`, found `()`
      |    |
      |    implicitly returns `()` as its body has no tail or `return` expression
      |
      = note:   expected type `!`
              found unit type `()`
  
  For more information about this error, try `rustc --explain E0308`.
  error: could not compile `simd-json` due to previous error
```